File: nanopdf-js/native/context.cc
/**
 * Context FFI Bindings
 *
 * Implements N-API bindings for MuPDF context functions (fz_context).
 * Context manages memory allocation, error handling, and resource tracking.
 */

#include <napi.h>
#include "include/mupdf.h"

/**
 * Create a new context
 * JavaScript: createContext(): NativeContext
 */
Napi::Value CreateContext(const Napi::CallbackInfo& info) {
    Napi::Env env = info.Env();
    
    fz_context ctx = fz_new_context(nullptr, nullptr, FZ_STORE_DEFAULT);
    if (ctx == 0) {
        Napi::Error::New(env, "Failed to create context").ThrowAsJavaScriptException();
        return env.Null();
    }
    
    Napi::Object obj = Napi::Object::New(env);
    obj.Set("_handle", Napi::Number::New(env, ctx));
    return obj;
}

/**
 * Drop/free a context
 * JavaScript: dropContext(ctx: NativeContext): void
 */
Napi::Value DropContext(const Napi::CallbackInfo& info) {
    Napi::Env env = info.Env();
    
    if (info.Length() < 1 || !info[0].IsObject()) {
        Napi::TypeError::New(env, "Expected context object").ThrowAsJavaScriptException();
        return env.Undefined();
    }
    
    Napi::Object ctxObj = info[0].As<Napi::Object>();
    if (!ctxObj.Has("_handle")) {
        Napi::TypeError::New(env, "Invalid context object").ThrowAsJavaScriptException();
        return env.Undefined();
    }
    
    fz_context ctx = ctxObj.Get("_handle").As<Napi::Number>().Int32Value();
    fz_drop_context(ctx);
    
    return env.Undefined();
}

/**
 * Clone a context (create a reference)
 * JavaScript: cloneContext(ctx: NativeContext): NativeContext
 */
Napi::Value CloneContext(const Napi::CallbackInfo& info) {
    Napi::Env env = info.Env();
    
    if (info.Length() < 1 || !info[0].IsObject()) {
        Napi::TypeError::New(env, "Expected context object").ThrowAsJavaScriptException();
        return env.Null();
    }
    
    Napi::Object ctxObj = info[0].As<Napi::Object>();
    fz_context ctx = ctxObj.Get("_handle").As<Napi::Number>().Int32Value();
    
    fz_context newCtx = fz_clone_context(ctx);
    
    Napi::Object obj = Napi::Object::New(env);
    obj.Set("_handle", Napi::Number::New(env, newCtx));
    return obj;
}

/**
 * Initialize context exports
 */
Napi::Object InitContext(Napi::Env env, Napi::Object exports) {
    exports.Set("createContext", Napi::Function::New(env, CreateContext));
    exports.Set("dropContext", Napi::Function::New(env, DropContext));
    exports.Set("cloneContext", Napi::Function::New(env, CloneContext));
    return exports;
}

File: nanopdf-js/deno/context.ts
/**
 * Context management for NanoPDF operations
 */

import { fz_new_context, fz_drop_context, fz_clone_context, FZ_STORE_DEFAULT } from "./ffi.ts";

export class Context {
  private handle: bigint;
  private dropped = false;

  constructor(maxStore: number = FZ_STORE_DEFAULT) {
    const handle = fz_new_context(null, null, maxStore);
    if (handle === 0n) {
      throw new Error("Failed to create context");
    }
    this.handle = handle;
  }

  getHandle(): bigint {
    if (this.dropped) {
      throw new Error("Context has been dropped");
    }
    return this.handle;
  }

  clone(): Context {
    const newHandle = fz_clone_context(this.handle);
    if (newHandle === 0n) {
      throw new Error("Failed to clone context");
    }
    const ctx = Object.create(Context.prototype);
    ctx.handle = newHandle;
    ctx.dropped = false;
    return ctx;
  }

  drop(): void {
    if (!this.dropped) {
      fz_drop_context(this.handle);
      this.dropped = true;
    }
  }

  [Symbol.dispose](): void {
    this.drop();
  }
}

